algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/GeneHasSufficientCopyNumber.kt

```diff
@@ -0,0 +1,106 @@
+
+    companion object {
+        fun fromCopyNumber(copyNumber: CopyNumber, requestedMinCopyNumber: Int): CopyNumberEvaluation {
+            return if (copyNumber.canonicalImpact.minCopies >= requestedMinCopyNumber) {
+                when {
+                    copyNumber.geneRole == GeneRole.TSG -> {
+                        SUFFICIENT_COPY_NUMBER_ON_NON_ONCOGENE
+                    }
+
+                    copyNumber.proteinEffect == ProteinEffect.LOSS_OF_FUNCTION
+                            || copyNumber.proteinEffect == ProteinEffect.LOSS_OF_FUNCTION_PREDICTED -> {
+                        SUFFICIENT_COPY_NUMBER_WITH_LOSS_OF_FUNCTION
+                    }
+
+                    !copyNumber.isReportable -> {
+                        UNREPORTABLE_SUFFICIENT_COPY_NUMBER
+                    }
+
+                    else -> {
+                        REPORTABLE_SUFFICIENT_COPY_NUMBER
+                    }
+                }
+            } else {
+                INSUFFICIENT_COPY_NUMBER
+            }
+        }
+    }
+}
+
+class GeneHasSufficientCopyNumber(private val gene: String, private val requestedMinCopyNumber: Int, maxTestAge: LocalDate? = null) :
+    MolecularEvaluationFunction(maxTestAge) {
+
+    override fun genes() = listOf(gene)
+
+    override fun evaluate(test: MolecularTest): Evaluation {
+        val evaluatedCopyNumbers: Map<CopyNumberEvaluation, Set<String>> = test.drivers.copyNumbers
+            .filter { it.gene == gene }
+            .groupingBy { CopyNumberEvaluation.fromCopyNumber(it, requestedMinCopyNumber) }
+            .fold(emptySet()) { acc, copyNumber -> acc + copyNumber.event }
+
+        val eventsOnNonCanonical = test.drivers.copyNumbers.filter { copyNumber ->
+            copyNumber.gene == gene && copyNumber.otherImpacts.any { it.minCopies >= requestedMinCopyNumber }
+        }.map { it.event }.toSet()
+
+        return when {
+            evaluatedCopyNumbers[CopyNumberEvaluation.REPORTABLE_SUFFICIENT_COPY_NUMBER] != null -> {
+                val reportableSufficientCN = evaluatedCopyNumbers[CopyNumberEvaluation.REPORTABLE_SUFFICIENT_COPY_NUMBER]!!
+                EvaluationFactory.pass("$gene copy number is above requested $requestedMinCopyNumber", inclusionEvents = reportableSufficientCN)
+            }
+
+            evaluatedCopyNumbers[CopyNumberEvaluation.UNREPORTABLE_SUFFICIENT_COPY_NUMBER] != null -> {
+                val unreportableSufficientCN = evaluatedCopyNumbers[CopyNumberEvaluation.UNREPORTABLE_SUFFICIENT_COPY_NUMBER]!!
+                EvaluationFactory.pass(
+                    "$gene has a copy number >$requestedMinCopyNumber copies but is considered not reportable",
+                    inclusionEvents = unreportableSufficientCN
+                )
+            }
+
+            else -> evaluatePotentialWarns(evaluatedCopyNumbers, eventsOnNonCanonical, test.evidenceSource)
+                ?: EvaluationFactory.fail("Insufficient copies of $gene")
+        }
```

Comment:
`when` is clear and consistent, but I just want to point out you could chain these together with elvis operators:
```suggestion
        return evaluatedCopyNumbers[CopyNumberEvaluation.REPORTABLE_SUFFICIENT_COPY_NUMBER]?.let { reportableSufficientCN ->
            EvaluationFactory.pass("$gene copy number is above requested $requestedMinCopyNumber", inclusionEvents = reportableSufficientCN)
        }
            ?: evaluatedCopyNumbers[CopyNumberEvaluation.UNREPORTABLE_SUFFICIENT_COPY_NUMBER]?.let { unreportableSufficientCN ->
                EvaluationFactory.pass(
                    "$gene has a copy number >$requestedMinCopyNumber copies but is considered not reportable",
                    inclusionEvents = unreportableSufficientCN
                )
            }
            ?: evaluatePotentialWarns(evaluatedCopyNumbers, eventsOnNonCanonical, test.evidenceSource)
            ?: EvaluationFactory.fail("Insufficient copies of $gene")
```

Reply:
Final suggestion - if we're sticking with `when`, it might be cleaner to define the CN selections first so you don't have to reference them multiple times and use `!!`:
```suggestion
        val reportableSufficientCN = evaluatedCopyNumbers[CopyNumberEvaluation.REPORTABLE_SUFFICIENT_COPY_NUMBER]
        val unreportableSufficientCN = evaluatedCopyNumbers[CopyNumberEvaluation.UNREPORTABLE_SUFFICIENT_COPY_NUMBER]
        return when {
            reportableSufficientCN != null -> {
                EvaluationFactory.pass("$gene copy number is above requested $requestedMinCopyNumber", inclusionEvents = reportableSufficientCN)
            }

            unreportableSufficientCN != null -> {
                EvaluationFactory.pass(
                    "$gene has a copy number >$requestedMinCopyNumber copies but is considered not reportable",
                    inclusionEvents = unreportableSufficientCN
                )
            }

            else -> evaluatePotentialWarns(evaluatedCopyNumbers, eventsOnNonCanonical, test.evidenceSource)
                ?: EvaluationFactory.fail("Insufficient copies of $gene")
        }
```
Final-final, alternative suggestion 🙂: if we keep having to do this for molecular rules, we could create another helper like https://github.com/hartwigmedical/actin/blob/78d308b608142743a5c9a2afc7dac4a651a9816a/algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/MolecularEventUtil.kt#L10 
for PASS cases.